### DISCUSSION TITLE
SelectConnection and BlockingConnection in same thread

### DISCUSSION THREAD

**Discussion by jmath222** [pika/pika#1466] [Q&A]
Are there any fundamental reasons why a SelectConnection and BlockingConnection shouldn't be used together in the same single-threaded application?

Our application uses the BlockingConnection to publish and the SelectConnection to consume. We're troubleshooting a rare issue where it appears the BlockingConnection blocks forever while trying to publish, with no evidence of the socket closing in either connection. I had a hunch that maybe the SelectConnection is somehow interfering, so I wanted to ask this question before troubleshooting further.

**Comment by lukebakken** [comment]
> Are there any fundamental reasons why a SelectConnection and BlockingConnection shouldn't be used together in the same single-threaded application?

Yes, absolutely.

Just FYI, `BlockingConnection` is built on `SelectConnection`.

An instance of `SelectConnection` will start an I/O loop on the current thread. There is no support for multiple I/O loops on the same thread.

You _must_ run each instance on its own thread. The fact that it works at all is kind of surprising to me.

**Reply by jmath222** [reply]
We only start one IO loop, the one from the SelectConnection, with a call to SelectConnection.ioloop.start(). Then the BlockingConnection is able to publish (except for this rare edge case) from within SelectConnection callbacks and timeouts.

It appears BlockingConnection has its own socket and poller so I'm not sure how SelectConnection could interfere while it's trying to publish, but can you think of a technical reason why this wouldn't work? I understand at this point it's bad practice, but updating our code to only use the SelectConnection would be a somewhat major change for us.

We've been running in production this way with pika 1.1.0 and 1.2.0 across multiple sites and RabbitMQ versions for nearly four years and have only experienced the issue 3-4 times. Although our use case doesn't involve a high volume of messaging.

**Reply by lukebakken** [reply]
> It appears BlockingConnection has its own socket and poller so I'm not sure how SelectConnection could interfere while it's trying to publish, but can you think of a technical reason why this wouldn't work?

In essence, you are calling `select` more than once in the same thread. It has been working only because of your low message rate and sheer luck.

Your use-case is not recommended, nor tested, nor supported.
